### PR TITLE
maintenanced: use before_shmem_exit to clear workerPid

### LINKED ISSUE #3751: Maintenance daemon crashed and couldn't start when a backend ran OOM
Logs were:

```
2020-04-08 13:26:50.0000000 2020-04-08 13:26:59.787 UTC [60765] FATAL:  out of memory
2020-04-08 13:26:59.787 UTC [60765] DETAIL:  Failed on request of size 41520 in memory context "WAL record construction".
2020-04-08 13:27:00.0000000 2020-04-08 13:27:01.036 UTC [60772] FATAL:  out of memory
2020-04-08 13:27:01.036 UTC [60772] DETAIL:  Failed while creating memory context "index info".
2020-04-08 13:27:00.0000000 2020-04-08 13:27:01.222 UTC [60773] FATAL:  out of memory
2020-04-08 13:27:01.222 UTC [60773] DETAIL:  Failed on request of size 41520 in memory context "WAL record construction".
2020-04-08 13:27:00.0000000 2020-04-08 13:27:01.683 UTC [60774] FATAL:  out of memory
2020-04-08 13:27:01.683 UTC [60774] DETAIL:  Failed on request of size 41520 in memory context "WAL record construction".
2020-04-08 13:27:00.0000000 2020-04-08 13:27:04.022 UTC [9] LOG:  background worker "Citus Maintenance Daemon: 16386 10" (PID 41620) exited with exit code 1
2020-04-08 13:27:00.0000000 2020-04-08 13:27:09.023 UTC [9] LOG:  could not fork worker process: Cannot allocate memory
```

Postgres tries to restart the worker process later, so I am not sure if we should be worried about this.

```c
/*
	* Try to start the worker.
	*
	* On failure, give up processing workers for now, but set
	* StartWorkerNeeded so we'll come back here on the next iteration
	* of ServerLoop to try again.  (We don't want to wait, because
	* there might be additional ready-to-run workers.)  We could set
	* HaveCrashedWorker as well, since this worker is now marked
	* crashed, but there's no need because the next run of this
	* function will do that.
	*/
if (!do_start_bgworker(rw))
{
	StartWorkerNeeded = true;
	return;
}
```

### PR DESCRIPTION
**Pull request by serprex**
*originally maintenanced: don't proc_exit(0) when myDbData->workerPid is not alive*

Tested by experimenting with raising an error at the end of the main loop with `rand`. This isn't a complete solution, `multi_extension` still times out so a maintenance daemon which crashes after each iteration of the main loop does lose progress. But we can at least have a new daemon take over

Fixes #3751 

*edit: this took a pivot, we now clean up entries so that workerPid should never be left pointing at a dead process*

### CODE REVIEW

**Comment by serprex** [comment]
This issue is when the maintenance daemon crashes such that there's no reset of shared memory, afaict `shmem_exit` isn't called in this scenario

**Comment by marcocitus** [comment]
If the any backend crashes then postgres restarts.

File: src/backend/distributed/utils/maintenanced.c [hunk 1]
```diff
@@ -267,9 +268,24 @@ CitusMaintenanceDaemonMain(Datum main_arg)
 		 * to properly add it to the hash.
 		 * Alternatively, don't continue if another worker exists.
 		 */
-		proc_exit(0);
+		bool immediatelyExit = true;
+		if (myDbData && myDbData->workerPid != 0)
+		{
```
Comment (marcocitus) [inline src/backend/distributed/utils/maintenanced.c:273]: I don't see a huge problem with this, but I'm wondering whether this is still relevant and might confuse future readers. 

The !myDbData case covers the restart-on-crash scenario where the shared memory is gone. 
The workerPid != 0 scenario used to cover restart-on-error scenarios, which are now handled by before_shmem_exit.

I don't believe there is a scenario in which `myDbData && myDbData->workerPid != 0` can be true now, or even a scenario in which `myDbData->workerPid != 0` can be true. However, I might be wrong.

I'm wondering whether we should change to something like:
```c
if (myDbData && myDbData->workerPid != 0 && kill(myDbData->workerPid, 0) != 0)
{
    /* 
     * The last worker process not longer exists, but failed to clean up. Set
     * daemonStarted to false to create a new daemon from the next session that
     * calls InitializeMaintenanceDaemonBackend.
     */
    myDbData->daemonStarted = false;
    myDbData->workerPid = 0;
}

proc_exit(0);
```

I'll leave it up to you to decide. I'm ok with all 3 options (reverting, leaving as is, or `daemonStarted = false`).

Comment (serprex) [reply src/backend/distributed/utils/maintenanced.c]: Removed `workerPid != 0` in favor of `Assert(myDbData->workerPid == 0)`

**Pull request merged**